File: src/main/java/code/trie/Trie.java
package code.trie;

public class Trie {

    TNode createAndGetNode() {
        TNode node = new TNode();
        node.isEOW = false;
        return node;
    }

    void insertWordInTrie(TNode root, String word) {
        for(char c : word.toCharArray()) {
            int a = (int) c;
            if(a > 64 && a < 91) a = a - 65;
            if(a > 96 && a < 122) a = a - 97;
            if(root.next[a] == null) root.next[a] = createAndGetNode();
            root = root.next[a];
         }
         root.isEOW = true;
    }

    void printWordsInTrieHelper(TNode root, char[] path, int pi) {
        if(root == null) return;
        if(root.isEOW) {
            for(int i=0;i<pi;i++) System.out.printf("%c", path[i]);
            System.out.printf(" ");
        }
        for(int i = 0; i < 26; i++) {
            path[pi] = (char) (i + 'a');
            printWordsInTrieHelper(root.next[i], path, pi+1);
        }
    }

    void printWordsInTrie(TNode root) { printWordsInTrieHelper(root, new char[101], 0); }

}

File: src/main/java/code/shubham/trie/Trie.go
package main
import "fmt"

type Node struct {
  var next map[byte]Node
  word string
}

func (n Node) add(word string) Node {
  return n.name
}


func solution(A []int) int {
    N := len(A);
    for i := 0; i < N; i++ {
        if A[i] <= 0 || A[i] >= N {
            continue;
        }
        cur := A[i];
        next := 0;
        for cur > 0 && cur <= N && A[cur - 1] != cur {
            next = A[cur - 1];
            A[cur - 1] = cur;
            cur = next;
        }
    }
    for i := 0; i < N; i++ {
        if A[i] != i+1 {
            return i + 1;
        }
    }
    return N + 1;
}

func main(){
	fmt.Println( solution([]int{1, 2, 3, 6 , 7}) );
}